contracts/exchange-wrappers/ZeroExExchangeWrapper.sol

```diff
@@ -13,38 +13,85 @@
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+/**
+ * @title ZeroExExchangeWrapper
+ * @author dYdX
+ *
+ * Wrapper around 0x API ERC20 token swap.
+ */
+contract ZeroExExchangeWrapper is I_ExchangeWrapper {
+    using SafeERC20 for IERC20;
 
-import { I_ExchangeWrapper } from "../external/I_ExchangeWrapper.sol";
-import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
 
-contract ZeroExExchangeWrapper is I_ExchangeWrapper  {
     // ============ Public Functions ============
 
     /**
-     * Exchange some amount of takerToken for makerToken. Just for tests
+     * Exchange some amount of takerToken for makerToken.
      *
      * @param  tradeOriginator      Address of the initiator of the trade (however, this value
      *                              cannot always be trusted as it is set at the discretion of the
      *                              msg.sender)
      * @param  receiver             Address to set allowance on once the trade has completed
      * @param  makerToken           Address of makerToken, the token to receive
      * @param  takerToken           Address of takerToken, the token to pay
-     * @param  inputTokenAmount     Amount of makerToken
+     * @param  requestedFillAmount  Amount of takerToken being paid
      * @param  orderData            Arbitrary bytes data for any information to pass to the exchange
      * @return                      The amount of makerToken received
      */
     function exchange(
         address tradeOriginator,
         address receiver,
-        ERC20 makerToken,
-        ERC20 takerToken,
-        uint256 inputTokenAmount,
+        IERC20 makerToken,
+        IERC20 takerToken,
+        uint256 requestedFillAmount,
         bytes calldata orderData
     )
         external override
         returns (uint256)
     {
-        // TODO implement a call to the actual ERC20 Transformation
-        return inputTokenAmount / uint(2);
+      address self = address(this);
+      address exchange = bytesToAddress(orderData);
+
+      uint256 originalTakerBalance = makerToken.balanceOf(self);
+
+      // Send fromToken to this contract.
+      IERC20(takerToken).safeTransferFrom(
+        msg.sender,
+        self,
+        requestedFillAmount
+      );
+
+      // Swap token
+      IERC20(takerToken).safeApprove(exchange, type(uint256).max);
+      (bool success, bytes memory returndata) = exchange.call(orderData[32:]);
+      require(success, string(returndata));
+
+      // find change in balance of takerToken on this contract
+      uint256 takerBalanceChange= IERC20(makerToken).balanceOf(self) - originalTakerBalance;
+
+      // transfer change in balance of takerToken to msg.sender
+      IERC20(makerToken).safeTransferFrom(
+        self,
+        msg.sender,
+        takerBalanceChange
+      );
```

Comment:
Just a heads up since `safeTransferFrom` passes control to token contract, we should keep reentrancy attacks in mind here... https://consensys.github.io/smart-contract-best-practices/known_attacks/